src/tree/gpu_hist/evaluate_splits.cu

```diff
@@ -8,316 +8,296 @@
-                                                &temp_storage);
+  GradientSumT parent_sum = lhs.computed_result.parent_sum;
+  GradientSumT left_sum, right_sum;
+  if (lhs.is_cat) {
+    left_sum = lhs.computed_result.parent_sum - rhs.gpair;
+    right_sum = rhs.gpair;
   } else {
-    EvaluateFeature<BLOCK_THREADS, SumReduceT, BlockScanT, MaxReduceT,
-                    TempStorage, GradientSumT,
-                    NumericBin<GradientSumT, TempStorage, BlockScanT>,
-                    UpdateNumeric<GradientSumT>>(fidx, inputs, evaluator, &best_split,
-                                                 &temp_storage);
+    if (forward) {
+      left_sum = lhs.computed_result.left_sum + rhs.gpair;
+      right_sum = lhs.computed_result.parent_sum - left_sum;
+    } else {
+      right_sum = lhs.computed_result.right_sum + rhs.gpair;
+      left_sum = lhs.computed_result.parent_sum - right_sum;
+    }
   }
+  bst_node_t nidx = (lhs.indicator == ChildNodeIndicator::kLeftChild) ? left.nidx : right.nidx;
+  float gain = evaluator.CalcSplitGain(
+      left.param, nidx, lhs.findex, GradStats{left_sum}, GradStats{right_sum});
+  float parent_gain = evaluator.CalcGain(left.nidx, left.param, GradStats{parent_sum});
+  float loss_chg = gain - parent_gain;
+  ret.computed_result = lhs.computed_result;
+  ret.computed_result.Update(left_sum, right_sum, parent_sum,
+                             loss_chg, lhs.findex, lhs.is_cat, lhs.fvalue,
+                             (forward ? DefaultDirection::kRightDir : DefaultDirection::kLeftDir),
+                             left.param);
+  return ret;
+}
 
-  cub::CTA_SYNC();
+template <typename GradientSumT>
+__noinline__ __device__ thrust::tuple<ScanElem<GradientSumT>, ScanElem<GradientSumT>>
+ScanOp<GradientSumT>::operator() (
+    thrust::tuple<ScanElem<GradientSumT>, ScanElem<GradientSumT>> lhs,
+    thrust::tuple<ScanElem<GradientSumT>, ScanElem<GradientSumT>> rhs) {
+  const auto& lhs_fw = thrust::get<0>(lhs);
+  const auto& lhs_bw = thrust::get<1>(lhs);
+  const auto& rhs_fw = thrust::get<0>(rhs);
+  const auto& rhs_bw = thrust::get<1>(rhs);
+  return thrust::make_tuple(DoIt<true>(lhs_fw, rhs_fw), DoIt<false>(lhs_bw, rhs_bw));
+};
 
-  if (threadIdx.x == 0) {
-    // Record best loss for each feature
-    out_candidates[blockIdx.x] = best_split;
+template <typename GradientSumT>
+template <bool forward>
+void
+__noinline__ __device__ WriteScan<GradientSumT>::DoIt(ScanElem<GradientSumT> e) {
+  EvaluateSplitInputs<GradientSumT>& split_input =
+      (e.indicator == ChildNodeIndicator::kLeftChild) ? left : right;
+  std::size_t offset = 0;
+  std::size_t n_features = left.feature_segments.empty() ? 0 : left.feature_segments.size() - 1;
+  if (e.indicator == ChildNodeIndicator::kRightChild) {
+    offset = n_features;
+  }
+  if ((!forward && split_input.feature_segments[e.findex] == e.hist_idx) ||
+      (forward && split_input.feature_segments[e.findex + 1] - 1 == e.hist_idx)) {
```

Comment:
I have a utility function in auc.cu that calculates the last index within the segment.